File: packages/python-sdk/tests/mock_build_api.py
"""In-memory mock of the template build control-plane API.

Mirrors packages/js-sdk/tests/template/mockBuildApi.ts: builds are simulated
synchronously at trigger time (one log entry per step) and drained through
status polling, the `base` alias is pre-seeded, and aliases/tags are tracked
in memory. The mock is installed by monkeypatching the build API functions
in the sync/async `main` modules (the same seam the stacktrace tests use),
so specialized fixtures can still override individual functions.
"""

from dataclasses import dataclass, field
from datetime import datetime
from types import SimpleNamespace
from typing import Dict, List, Optional, Tuple
from uuid import uuid4

from e2b.exceptions import BuildException, TemplateException
from e2b.template.logger import LogEntry
from e2b.template.types import (
    BuildStatusReason,
    InstructionType,
    TemplateBuildStatus,
    TemplateBuildStatusResponse,
    TemplateTag,
    TemplateTagInfo,
)

# Users that exist in the base image, like on the real build backend.
_VALID_USERS = {"root", "user"}


@dataclass
class MockBuild:
    template_id: str
    build_id: str
    alias: str
    tags: List[str]
    triggered: bool
    log_entries: List[LogEntry] = field(default_factory=list)
    final_status: TemplateBuildStatus = TemplateBuildStatus.READY
    reason: Optional[BuildStatusReason] = None
    created_at: datetime = field(default_factory=datetime.now)


def _log_entry(message: str) -> LogEntry:
    return LogEntry(timestamp=datetime.now(), level="info", message=message)


def _split_name(name: str) -> Optional[Tuple[str, Optional[str]]]:
    if ":" not in name:
        return (name, None) if name else None
    alias, _, tag = name.partition(":")
    if not alias or not tag:
        return None
    return (alias, tag)


class MockBuildAPI:
    def __init__(self):
        # Builds keyed by build ID; template IDs keyed by alias.
        self.builds: Dict[str, MockBuild] = {}
        self.templates: Dict[str, str] = {}
        self._seed_template("base")

    def _seed_template(self, alias: str) -> None:
        build = MockBuild(
            template_id=str(uuid4()),
            build_id=str(uuid4()),
            alias=alias,
            tags=["latest"],
            triggered=True,
            log_entries=[_log_entry("Build finished")],
        )
        self.builds[build.build_id] = build
        self.templates[alias] = build.template_id

    def _latest_build_for_alias(self, alias: str) -> Optional[MockBuild]:
        latest = None
        for build in self.builds.values():
            if build.alias == alias:
                latest = build
        return latest

    def request_build(self, client, name, tags, cpu_count, memory_mb):
        parsed = _split_name(name)
        if parsed is None:
            raise BuildException(f"Invalid template name: '{name}'")

        alias, tag = parsed
        template_id = self.templates.setdefault(alias, str(uuid4()))

        all_tags = ([tag] if tag else []) + (tags or [])
        build = MockBuild(
            template_id=template_id,
            build_id=str(uuid4()),
            alias=alias,
            tags=all_tags,
            triggered=False,
        )
        self.builds[build.build_id] = build

        return SimpleNamespace(
            template_id=template_id, build_id=build.build_id, tags=all_tags
        )

    def get_file_upload_link(self, client, template_id, files_hash, stack_trace=None):
        # Reporting every hash as cached (with no upload URL) skips uploads.
        return SimpleNamespace(present=True, url=None)

    def trigger_build(self, client, template_id, build_id, template) -> None:
        build = self.builds.get(build_id)
        if build is None or build.template_id != template_id:
            raise BuildException("Build not found")

        # Simulate the build synchronously: one log entry per step.
        from_value = template.get("fromImage") or template.get("fromTemplate") or "base"
        build.log_entries.append(_log_entry(f"FROM {from_value}"))
        for index, step in enumerate(template.get("steps") or []):
            step_type = InstructionType(step.get("type")).value
            args = step.get("args") or []
            # RUN steps carry the user in args[1]; only users that exist in
            # the base image are accepted, like the real build backend.
            user = args[1] if step_type == "RUN" and len(args) > 1 else None
            if user is not None and user not in _VALID_USERS:
                build.final_status = TemplateBuildStatus.ERROR
                build.reason = BuildStatusReason(
                    message=f"failed to run command '{args[0]}': command failed: "
                    f"unauthenticated: invalid username: '{user}'",
                    step=str(index + 1),
                )
                break
            build.log_entries.append(
                _log_entry(f"Step {index + 1}: {step_type} {' '.join(args)}")
            )
        if build.final_status != TemplateBuildStatus.ERROR:
            build.log_entries.append(_log_entry("Build finished"))
        build.triggered = True

    def get_build_status(
        self, client, template_id, build_id, logs_offset
    ) -> TemplateBuildStatusResponse:
        build = self.builds.get(build_id)
        if build is None or build.template_id != template_id:
            raise BuildException("Build not found")

        log_entries = build.log_entries[logs_offset:]

        # Deliver the pending log entries first (status `building`), then
        # report the final status once they are drained.
        if not build.triggered:
            status = TemplateBuildStatus.WAITING
        elif log_entries:
            status = TemplateBuildStatus.BUILDING
        else:
            status = build.final_status

        return TemplateBuildStatusResponse(
            build_id=build.build_id,
            template_id=build.template_id,
            status=status,
            log_entries=log_entries,
            logs=[entry.message for entry in log_entries],
            reason=build.reason if status == TemplateBuildStatus.ERROR else None,
        )

    def check_alias_exists(self, client, alias) -> bool:
        return alias in self.templates

    def assign_tags(self, client, target_name, tags) -> TemplateTagInfo:
        parsed = _split_name(target_name)
        if parsed is None:
            raise TemplateException(f"Invalid target: '{target_name}'")

        build = self._latest_build_for_alias(parsed[0])
        if build is None:
            raise TemplateException("Template not found")

        # Tags may be bare ('production') or namespaced ('alias:production');
        # the API returns and stores just the tag portion.
        assigned = []
        for tag in tags:
            parsed_tag = _split_name(tag)
            if parsed_tag is None:
                raise TemplateException(f"Invalid tag: '{tag}'")
            assigned.append(parsed_tag[1] or parsed_tag[0])

        build.tags.extend(assigned)

        return TemplateTagInfo(build_id=build.build_id, tags=assigned)

    def remove_tags(self, client, name, tags) -> None:
        build = self._latest_build_for_alias(name)
        if build is None:
            raise TemplateException("Template not found")

        build.tags = [tag for tag in build.tags if tag not in tags]

    def get_template_tags(self, client, template_id_or_name) -> List[TemplateTag]:
        template_builds = [
            build
            for build in self.builds.values()
            if build.template_id == template_id_or_name
        ]
        if not template_builds:
            raise TemplateException("Template not found")

        return [
            TemplateTag(tag=tag, build_id=build.build_id, created_at=build.created_at)
            for build in template_builds
            for tag in build.tags
        ]

    _SYNC_FUNCTIONS = (
        "request_build",
        "get_file_upload_link",
        "trigger_build",
        "get_build_status",
        "check_alias_exists",
        "assign_tags",
        "remove_tags",
        "get_template_tags",
    )

    def install_sync(self, monkeypatch) -> None:
        import e2b.template_sync.build_api as build_api_mod
        import e2b.template_sync.main as main_mod

        for name in self._SYNC_FUNCTIONS:
            monkeypatch.setattr(main_mod, name, getattr(self, name))
        # wait_for_build_finish polls get_build_status through its own module.
        monkeypatch.setattr(build_api_mod, "get_build_status", self.get_build_status)

    def install_async(self, monkeypatch) -> None:
        import e2b.template_async.build_api as build_api_mod
        import e2b.template_async.main as main_mod

        def as_async(func):
            async def wrapper(*args, **kwargs):
                return func(*args, **kwargs)

            return wrapper

        for name in self._SYNC_FUNCTIONS:
            monkeypatch.setattr(main_mod, name, as_async(getattr(self, name)))
        # wait_for_build_finish polls get_build_status through its own module.
        monkeypatch.setattr(
            build_api_mod, "get_build_status", as_async(self.get_build_status)
        )

File: packages/js-sdk/tests/template/mockBuildApi.ts
import { randomUUID } from 'node:crypto'
import { http, HttpResponse, type HttpHandler } from 'msw'

import { apiUrl } from '../setup'

interface MockLogEntry {
  timestamp: string
  level: 'debug' | 'info' | 'warn' | 'error'
  message: string
}

interface MockBuild {
  templateID: string
  buildID: string
  alias: string
  tags: string[]
  triggered: boolean
  logEntries: MockLogEntry[]
  finalStatus: 'ready' | 'error'
  reason?: { message: string; step?: string }
  createdAt: string
}

interface MockBuildApiState {
  /** Builds keyed by buildID. */
  builds: Map<string, MockBuild>
  /** Template IDs keyed by alias. */
  templates: Map<string, string>
}

const VALID_USERS = new Set(['root', 'user'])

function logEntry(message: string): MockLogEntry {
  return { timestamp: new Date().toISOString(), level: 'info', message }
}

function splitName(name: string): { alias: string; tag?: string } | undefined {
  const colonIndex = name.indexOf(':')
  if (colonIndex === -1) {
    return name.length > 0 ? { alias: name } : undefined
  }
  const alias = name.slice(0, colonIndex)
  const tag = name.slice(colonIndex + 1)
  if (alias.length === 0 || tag.length === 0) {
    return undefined
  }
  return { alias, tag }
}

function badRequest(message: string) {
  return HttpResponse.json({ code: 400, message }, { status: 400 })
}

function notFound(message: string) {
  return HttpResponse.json({ code: 404, message }, { status: 404 })
}

/**
 * Stateful in-memory mock of the template build API (request build, file
 * upload links, trigger, status polling, aliases, and tags). The `base`
 * alias is pre-seeded so `fromTemplate('base')` and `Template.exists('base')`
 * work out of the box.
 */
export function createMockBuildApi(): {
  handlers: HttpHandler[]
  state: MockBuildApiState
} {
  const builds = new Map<string, MockBuild>()
  const templates = new Map<string, string>()

  function seedTemplate(alias: string) {
    const build: MockBuild = {
      templateID: randomUUID(),
      buildID: randomUUID(),
      alias,
      tags: ['latest'],
      triggered: true,
      logEntries: [logEntry('Build finished')],
      finalStatus: 'ready',
      createdAt: new Date().toISOString(),
    }
    builds.set(build.buildID, build)
    templates.set(alias, build.templateID)
  }

  seedTemplate('base')

  function latestBuildForAlias(alias: string): MockBuild | undefined {
    let latest: MockBuild | undefined
    for (const build of builds.values()) {
      if (build.alias === alias) {
        latest = build
      }
    }
    return latest
  }

  const handlers = [
    // Request a template build
    http.post(apiUrl('/v3/templates'), async ({ request }) => {
      const body = (await request.clone().json()) as {
        name: string
        tags?: string[]
      }

      const parsedName = splitName(body.name)
      if (!parsedName) {
        return badRequest(`Invalid template name: '${body.name}'`)
      }

      const { alias, tag } = parsedName
      const templateID = templates.get(alias) ?? randomUUID()
      templates.set(alias, templateID)

      const tags = [...(tag ? [tag] : []), ...(body.tags ?? [])]
      const build: MockBuild = {
        templateID,
        buildID: randomUUID(),
        alias,
        tags,
        triggered: false,
        logEntries: [],
        finalStatus: 'ready',
        createdAt: new Date().toISOString(),
      }
      builds.set(build.buildID, build)

      return HttpResponse.json({
        templateID,
        buildID: build.buildID,
        public: false,
        names: [alias],
        tags,
        aliases: [alias],
      })
    }),

    // Check whether the files for a hash are already uploaded. Always
    // reporting them as cached (with no upload URL) skips the upload step.
    http.get(apiUrl('/templates/:templateID/files/:hash'), () => {
      return HttpResponse.json({ present: true })
    }),

    // Trigger a build: simulate it synchronously by recording one log entry
    // per step, then mark the final status.
    http.post<{ templateID: string; buildID: string }>(
      apiUrl('/v2/templates/:templateID/builds/:buildID'),
      async ({ params, request }) => {
        const build = builds.get(params.buildID)
        if (!build || build.templateID !== params.templateID) {
          return notFound('Build not found')
        }

        const body = (await request.clone().json()) as {
          fromImage?: string
          fromTemplate?: string
          steps?: { type: string; args?: string[] }[]
        }

        const from = body.fromImage ?? body.fromTemplate ?? 'base'
        build.logEntries.push(logEntry(`FROM ${from}`))
        for (const [index, step] of (body.steps ?? []).entries()) {
          // RUN steps carry the user in args[1]; only users that exist in the
          // base image are accepted, like the real build backend.
          const user = step.type === 'RUN' ? step.args?.[1] : undefined
          if (user && !VALID_USERS.has(user)) {
            build.finalStatus = 'error'
            build.reason = {
              message: `failed to run command '${step.args?.[0]}': command failed: unauthenticated: invalid username: '${user}'`,
              step: String(index + 1),
            }
            break
          }
          build.logEntries.push(
            logEntry(
              `Step ${index + 1}: ${step.type} ${(step.args ?? []).join(' ')}`
            )
          )
        }
        if (build.finalStatus !== 'error') {
          build.logEntries.push(logEntry('Build finished'))
        }
        build.triggered = true

        return HttpResponse.json({}, { status: 202 })
      }
    ),

    // Poll build status: deliver the pending log entries on the first call
    // (status `building`), then report the final status once drained.
    http.get<{ templateID: string; buildID: string }>(
      apiUrl('/templates/:templateID/builds/:buildID/status'),
      ({ params, request }) => {
        const build = builds.get(params.buildID)
        if (!build || build.templateID !== params.templateID) {
          return notFound('Build not found')
        }

        const logsOffset = Number(
          new URL(request.url).searchParams.get('logsOffset') ?? 0
        )
        const logEntries = build.logEntries.slice(logsOffset)

        let status: string
        if (!build.triggered) {
          status = 'waiting'
        } else if (logEntries.length > 0) {
          status = 'building'
        } else {
          status = build.finalStatus
        }

        return HttpResponse.json({
          templateID: build.templateID,
          buildID: build.buildID,
          status,
          logEntries,
          logs: logEntries.map((entry) => entry.message),
          ...(status === 'error' ? { reason: build.reason } : {}),
        })
      }
    ),

    // Check whether an alias exists
    http.get<{ alias: string }>(
      apiUrl('/templates/aliases/:alias'),
      ({ params }) => {
        const templateID = templates.get(params.alias)
        if (!templateID) {
          return notFound('Template not found')
        }
        return HttpResponse.json({ templateID, public: false })
      }
    ),

    // Assign tags to an existing build
    http.post(apiUrl('/templates/tags'), async ({ request }) => {
      const body = (await request.clone().json()) as {
        target: string
        tags: string[]
      }

      const parsedTarget = splitName(body.target)
      if (!parsedTarget) {
        return badRequest(`Invalid target: '${body.target}'`)
      }

      const build = latestBuildForAlias(parsedTarget.alias)
      if (!build) {
        return notFound('Template not found')
      }

      // Tags may be bare ('production') or namespaced ('alias:production');
      // the API returns and stores just the tag portion.
      const tags: string[] = []
      for (const tag of body.tags) {
        const parsedTag = splitName(tag)
        if (!parsedTag) {
          return badRequest(`Invalid tag: '${tag}'`)
        }
        tags.push(parsedTag.tag ?? parsedTag.alias)
      }

      build.tags.push(...tags)

      return HttpResponse.json({ buildID: build.buildID, tags })
    }),

    // Remove tags from a template
    http.delete(apiUrl('/templates/tags'), async ({ request }) => {
      const body = (await request.clone().json()) as {
        name: string
        tags: string[]
      }

      const build = latestBuildForAlias(body.name)
      if (!build) {
        return notFound('Template not found')
      }

      build.tags = build.tags.filter((tag) => !body.tags.includes(tag))

      return new HttpResponse(null, { status: 204 })
    }),

    // List tags for a template
    http.get<{ templateID: string }>(
      apiUrl('/templates/:templateID/tags'),
      ({ params }) => {
        const templateBuilds = Array.from(builds.values()).filter(
          (build) => build.templateID === params.templateID
        )
        if (templateBuilds.length === 0) {
          return notFound('Template not found')
        }

        return HttpResponse.json(
          templateBuilds.flatMap((build) =>
            build.tags.map((tag) => ({
              tag,
              buildID: build.buildID,
              createdAt: build.createdAt,
            }))
          )
        )
      }
    ),
  ]

  return { handlers, state: { builds, templates } }
}
